overlord/snapstate/autorefresh.go

```diff
@@ -402,44 +384,48 @@ func (m *autoRefresh) Ensure() (err error) {
 	return err
 }
 
-func canContinueAutoRefresh(st *state.State) (int, bool) {
-	if cachedAttempt := st.Cached("auto-refresh-continue-attempt"); cachedAttempt != nil {
-		return cachedAttempt.(int), true
+func (m *autoRefresh) restoreMonitoring() error {
+	if m.restoredMonitoring {
+		return nil
 	}
 
-	return 0, false
-}
+	// clear the old monitoring state; remove in a later snapd version
+	m.state.Set("monitored-snaps", nil)
 
-func (m *autoRefresh) restoreMonitoring() {
-	if m.restoredMonitoring {
-		return
+	var refreshHints map[string]*refreshCandidate
+	if err := m.state.Get("refresh-candidates", &refreshHints); err != nil && !errors.Is(err, &state.NoStateError{}) {
+		return fmt.Errorf("cannot get refresh-candidates: %v", err)
 	}
 
-	var monitored []string
-	if err := m.state.Get("monitored-snaps", &monitored); err != nil && !errors.Is(err, state.ErrNoState) {
-		logger.Noticef("cannot restore monitoring: %v", err)
-		return
+	defer func() { m.restoredMonitoring = true }()
+
+	var monitored []*SnapSetup
+	for _, hint := range refreshHints {
+		if hint.Monitoring {
+			monitored = append(monitored, &hint.SnapSetup)
+		}
 	}
 
-	defer func() { m.restoredMonitoring = true }()
 	if len(monitored) == 0 {
-		return
+		return nil
 	}
 
-	monitoring := make(map[string]chan<- bool)
+	abortChans := make(map[string]chan<- bool, len(monitored))
 	for _, snap := range monitored {
 		done := make(chan string, 1)
-		if err := cgroupMonitorSnapEnded(snap, done); err != nil {
-			logger.Noticef("cannot restore monitoring for snap %q closure: %v", snap, err)
+		snapName := snap.InstanceName()
+		if err := cgroupMonitorSnapEnded(snapName, done); err != nil {
+			logger.Noticef("cannot restore monitoring for snap %q closure: %v", snapName, err)
 			continue
 		}
 
 		abort := make(chan bool, 1)
-		monitoring[snap] = abort
-
+		abortChans[snapName] = abort
 		go continueRefreshOnSnapClose(m.state, snap, done, abort)
 	}
-	updateMonitoringState(m.state, monitoring)
```

Comment:
have we removed this helper now?

Reply:
Yes, I simplified the usage of the helpers a bit. The old version return and accepted the whole map of snaps to channels (possibly nil to remove) and then computed the things to put in the state and cache based on that. The new version splits adding and removing and passes less information around: `addMonitoring(st, snapName, abortChan)` and `removeMonitoring(st, snapName)`